File: hit.py
from globals import *

__all__ = ['monster_hit', 'rogue_hit', 'monster_damage', 'fight', 'get_dir_rc', 'get_hit_chance', 'get_weapon_damage']

def monster_hit(monster, other):
    if g.fight_monster and monster != g.fight_monster:
        g.fight_monster = None
    monster.trow = -1
    hit_chance = monster.clasz
    hit_chance -= rogue.exp + rogue.exp
    if hit_chance < 0: hit_chance = 0
    
    if not g.fight_monster:
        g.interrupted = 1
    
    mn = monster_name(monster)
    
    if not rand_percent(hit_chance):
        if not g.fight_monster:
            g.hit_message += "the %s misses" % (other if other else mn)
            message(g.hit_message, 0)
            g.hit_message = ""
        return
    
    if not g.fight_monster:
        g.hit_message += "the %s hit" % (other if other else mn)
        message(g.hit_message, 0)
        g.hit_message = ""
    
    if monster.ichar != 'F':
        damage = get_damage(monster.damage, 1)
        minus = (get_armor_class(rogue.armor) * 3.0) / 100.0 * damage
        damage -= int(minus)
    else:
        damage = monster.identified
        monster.identified += 1
    
    if damage > 0:
        rogue_damage(damage, monster)
    
    special_hit(monster)

def rogue_hit(monster):
    if check_xeroc(monster):
        return
    hit_chance = get_hit_chance(rogue.weapon)
    if not rand_percent(hit_chance):
        if not g.fight_monster:
            g.hit_message = "you miss  "
        #goto RET
        check_orc(monster)
        wake_up(monster)
        return
    
    damage = get_weapon_damage(rogue.weapon)
    if monster_damage(monster, damage): # still alive?
        if not g.fight_monster:
            g.hit_message = "you hit  "
    #RET:
    check_orc(monster)
    wake_up(monster)

def rogue_damage(d, monster):
    if d >= rogue.hp_current:
        rogue.hp_current = 0
        print_stats()
        killed_by(monster, 0)
    rogue.hp_current -= d
    print_stats()

def get_damage(ds, r):
    total = 0
    i = 0
    while i < len(ds):
        n = get_number(ds[i:])
        while i < len(ds) and ds[i] != 'd':
            i += 1
        i += 1
        d = get_number(ds[i:])
        while i < len(ds) and ds[i] != '/':
            i += 1
        for j in range(n):
            if r:
                total += get_rand(1, d)
            else:
                total += d
        if i < len(ds) and ds[i] == '/':
            i += 1
    return total

def get_w_damage(obj):
    if not obj:
        return -1
    to_hit = get_number(obj.damage) + obj.to_hit_enchantment
    i = 0
    while i < len(obj.damage) and obj.damage[i] != 'd':
        i += 1
    i += 1
    damage = get_number(obj.damage[i:]) + obj.damage_enchantment
    
    return get_damage("%dd%d" % (to_hit, damage), 1)

def get_number(s):
    total = 0
    i = 0
    while i < len(s) and '0' <= s[i] <= '9':
        total = 10 * total + ord(s[i]) - ord('0')
        i += 1
    return total

def to_hit(obj):
    if not obj:
        return 1
    return get_number(obj.damage) + obj.to_hit_enchantment

def damage_for_strength(s):
    if s <= 6: return s - 5
    if s <= 14: return 1
    if s <= 17: return 3
    if s <= 18: return 4
    if s <= 20: return 5
    if s <= 21: return 6
    if s <= 30: return 7
    return 8

def monster_damage(monster, damage):
    monster.quantity -= damage
    if monster.quantity <= 0:
        row = monster.row
        col = monster.col
        remove_mask(row, col, MONSTER)
        mvaddch(row, col, get_room_char(screen[row][col], row, col))
        refresh()
        
        g.fight_monster = None
        cough_up(monster)
        g.hit_message += "defeated the %s" % monster_name(monster)
        message(g.hit_message, 1)
        g.hit_message = ""
        add_exp(monster.kill_exp)
        print_stats()
        remove_from_pack(monster, g.level_monsters)
        
        if monster.ichar == 'F':
            g.being_held = 0
        
        return 0
    return 1

def fight(to_the_death):
    first_miss = 1
    ch = getchar()
    while not is_direction(ch):
        beep()
        if first_miss:
            message("direction?", 0)
            first_miss = 0
        ch = getchar()
    check_message()
    if ch == CANCEL:
        return
    
    row, col = get_dir_rc(ch, rogue.row, rogue.col)
    
    if not screen[row][col] & MONSTER or g.blind or hiding_xeroc(row, col):
        #MN:
        message("I see no monster there", 0)
        return
    g.fight_monster = object_at(g.level_monsters, row, col)
    if g.fight_monster.m_flags & IS_INVIS and not g.detect_monster:
        #goto MN
        message("I see no monster there", 0)
        return
    possible_damage = get_damage(g.fight_monster.damage, 0) * 2 / 3
    
    while g.fight_monster:
        single_move_rogue(ch, 0)
        if not to_the_death and rogue.hp_current <= possible_damage:
            g.fight_monster = None
        if not screen[row][col] & MONSTER or g.interrupted:
            g.fight_monster = None

def get_dir_rc(dir, row, col):
    if dir in "hyb":
        if col > 0: col -= 1
    if dir in "jnb":
        if row < LINES - 2: row += 1
    if dir in 'kyu':
        if row > MIN_ROW: row -= 1
    if dir in "lun":
        if col < COLS - 1: col += 1
    return row, col

def get_hit_chance(weapon):
    hit_chance = 40
    hit_chance += 3 * to_hit(weapon)
    hit_chance += (rogue.exp + rogue.exp)
    if hit_chance > 100: hit_chance = 100
    return hit_chance

def get_weapon_damage(weapon):
    damage = get_w_damage(weapon)
    damage += damage_for_strength(rogue.strength_current)
    damage += (rogue.exp + 1) / 2
    return damage

File: flrogue/lib/rogue/hit.dart
import 'globals.dart';
import 'level.dart';
import 'message.dart';
import 'monster.dart';
import 'move.dart';
import 'object.dart';
import 'pack.dart';
import 'room.dart';
import 'score.dart';
import 'special_hit.dart';
import 'ui.dart';

// Check if a character is a direction
bool isDirection(String ch) {
  return ch == 'h' ||
      ch == 'j' ||
      ch == 'k' ||
      ch == 'l' ||
      ch == 'y' ||
      ch == 'u' ||
      ch == 'n' ||
      ch == 'b' ||
      ch == cancel;
}

Future<void> monsterHit(GameObject monster, String? other) async {
  if (fightMonster != null && monster != fightMonster) {
    fightMonster = null;
  }
  monster.trow = -1;

  int hitChance = monster.clasz;
  hitChance -= rogue.exp + rogue.exp;
  if (hitChance < 0) hitChance = 0;

  if (fightMonster == null) {
    interrupted = true;
  }

  String mn = monsterName(monster);

  if (!randPercent(hitChance)) {
    if (fightMonster == null) {
      hitMessage += "the ${other ?? mn} misses";
      await message(hitMessage);
      hitMessage = "";
    }
    return;
  }

  if (fightMonster == null) {
    hitMessage += "the ${other ?? mn} hit";
    await message(hitMessage);
    hitMessage = "";
  }

  int damage;
  if (monster.ichar != 'F') {
    damage = _getDamage(monster.damage, true);
    double minus = (getArmorClass(rogue.armor) * 3.0) / 100.0 * damage;
    damage -= minus.toInt();
  } else {
    damage = monster.identified;
    monster.identified += 1;
  }

  if (damage > 0) {
    await _rogueDamage(damage, monster);
  }

  await specialHit(monster);
}

Future<void> rogueHit(GameObject monster) async {
  if (await checkXeroc(monster)) {
    return;
  }

  int hitChance = getHitChance(rogue.weapon);
  if (!randPercent(hitChance)) {
    if (fightMonster == null) {
      hitMessage = "you miss  ";
    }
    checkOrc(monster);
    wakeUp(monster);
    return;
  }

  int damage = getWeaponDamage(rogue.weapon);
  if (await monsterDamage(monster, damage)) {
    // still alive?
    if (fightMonster == null) {
      hitMessage = "you hit  ";
    }
  }

  checkOrc(monster);
  wakeUp(monster);
}

Future<void> _rogueDamage(int d, GameObject monster) async {
  if (d >= rogue.hpCurrent) {
    rogue.hpCurrent = 0;
    printStats();
    await killedBy(monster, DeathCause.monster);
  }
  rogue.hpCurrent -= d;
  printStats();
}

int _getDamage(String ds, bool r) {
  int total = 0;
  int i = 0;

  while (i < ds.length) {
    int n = _getNumber(ds.substring(i));
    while (i < ds.length && ds[i] != 'd') {
      i++;
    }
    i++;
    int d = _getNumber(ds.substring(i));
    while (i < ds.length && ds[i] != '/') {
      i++;
    }
    for (int j = 0; j < n; j++) {
      if (r) {
        total += getRand(1, d);
      } else {
        total += d;
      }
    }
    if (i < ds.length && ds[i] == '/') {
      i++;
    }
  }
  return total;
}

int _getWDamage(GameObject? obj) {
  if (obj == null) {
    return -1;
  }

  int toHit = _getNumber(obj.damage) + obj.toHitEnchantment;
  int i = 0;

  while (i < obj.damage.length && obj.damage[i] != 'd') {
    i++;
  }
  i++;

  int damage = _getNumber(obj.damage.substring(i)) + obj.damageEnchantment;

  return _getDamage("${toHit}d$damage", true);
}

int _getNumber(String s) {
  int total = 0;
  int i = 0;

  while (i < s.length && s[i].between('0', '9')) {
    total = 10 * total + (s[i].ascii - '0'.ascii);
    i++;
  }

  return total;
}

int _toHit(GameObject? obj) {
  if (obj == null) {
    return 1;
  }
  return _getNumber(obj.damage) + obj.toHitEnchantment;
}

int _damageForStrength(int s) {
  if (s <= 6) return s - 5;
  if (s <= 14) return 1;
  if (s <= 17) return 3;
  if (s <= 18) return 4;
  if (s <= 20) return 5;
  if (s <= 21) return 6;
  if (s <= 30) return 7;
  return 8;
}

Future<bool> monsterDamage(GameObject monster, int damage) async {
  monster.quantity -= damage;

  if (monster.quantity <= 0) {
    int row = monster.row;
    int col = monster.col;
    removeMask(row, col, Cell.monster);
    ui.move(row, col).write(getRoomChar(screen[row][col], row, col));
    ui.refresh();

    fightMonster = null;
    coughUp(monster);
    hitMessage += "defeated the ${monsterName(monster)}";
    await message(hitMessage, true);
    hitMessage = "";
    await addExp(monster.killExp);
    printStats();
    removeFromPack(monster, levelMonsters);

    if (monster.ichar == 'F') {
      beingHeld = false;
    }

    return false;
  }
  return true;
}

Future<void> fight(bool toTheDeath) async {
  bool firstMiss = true;
  String ch = await ui.getchar();

  while (!isDirection(ch)) {
    ui.beep();
    if (firstMiss) {
      await message("direction?");
      firstMiss = false;
    }
    ch = await ui.getchar();
  }

  checkMessage();
  if (ch == cancel) {
    return;
  }

  final (row, col) = getDirRc(ch, rogue.row, rogue.col);

  if (!(screen[row][col] & Cell.monster != 0) ||
      blind != 0 ||
      hidingXeroc(row, col)) {
    await message("I see no monster there");
    return;
  }

  fightMonster = objectAt(levelMonsters, row, col);
  if (fightMonster!.flagsIs(MonsterFlags.isInvis) && !detectMonster) {
    await message("I see no monster there");
    return;
  }

  int possibleDamage = _getDamage(fightMonster!.damage, false) * 2 ~/ 3;

  while (fightMonster != null) {
    await singleMoveRogue(ch, false);
    if (!toTheDeath && rogue.hpCurrent <= possibleDamage) {
      fightMonster = null;
    }
    if (!(screen[row][col] & Cell.monster != 0) || interrupted) {
      fightMonster = null;
    }
  }
}

(int, int) getDirRc(String dir, int row, int col) {
  if (dir == 'h' || dir == 'y' || dir == 'b') {
    if (col > 0) col -= 1;
  }
  if (dir == 'j' || dir == 'n' || dir == 'b') {
    if (row < ui.rows - 2) row += 1;
  }
  if (dir == 'k' || dir == 'y' || dir == 'u') {
    if (row > minRow) row -= 1;
  }
  if (dir == 'l' || dir == 'u' || dir == 'n') {
    if (col < ui.cols - 1) col += 1;
  }
  return (row, col);
}

int getHitChance(GameObject? weapon) {
  int hitChance = 40;
  hitChance += 3 * _toHit(weapon);
  hitChance += (rogue.exp + rogue.exp);
  if (hitChance > 100) hitChance = 100;
  return hitChance;
}

int getWeaponDamage(GameObject? weapon) {
  int damage = _getWDamage(weapon);
  damage += _damageForStrength(rogue.strengthCurrent);
  damage += (rogue.exp + 1) ~/ 2;
  return damage;
}
